blob/azureblob/azureblob.go

```diff
@@ -0,0 +1,570 @@
+	var err error
+	srcBlobParts.SAS, err = azblob.BlobSASSignatureValues{
+		Protocol:      azblob.SASProtocolHTTPS,
+		ExpiryTime:    time.Now().UTC().Add(opts.Expiry),
+		ContainerName: b.name,
+		BlobName:      key,
+		Permissions:   azblob.BlobSASPermissions{Add: true, Create: true, Delete: true, Read: true, Write: true}.String(),
+	}.NewSASQueryParameters(b.credential)
+
+	if err != nil {
+		return "", err
+	}
+
+	srcBlobURLWithSAS := srcBlobParts.URL()
+	return srcBlobURLWithSAS.String(), nil
+}
+
+type writer struct {
+	ctx         context.Context
+	urls        *serviceUrls
+	key         string
+	contentType string
+	opts        *driver.WriterOptions
+
+	w     *io.PipeWriter
+	donec chan struct{}
+	err   error
+}
+
+// NewTypedWriter implements driver.NewTypedWriter.
+func (b *bucket) NewTypedWriter(ctx context.Context, key string, contentType string, opts *driver.WriterOptions) (driver.Writer, error) {
+	containerURL := b.urls.serviceURL.NewContainerURL(b.name)
+	key = strings.Replace(key, OSPathSeparator, b.defaultDelimiter, -1)
+	blockBlobURL := containerURL.NewBlockBlobURL(key)
+
+	if opts.Metadata == nil {
+		opts.Metadata = map[string]string{}
+	}
+	if opts.BufferSize == 0 {
+		opts.BufferSize = defaultUploadBlockSize
+	}
+
+	w := &writer{
+		ctx:         ctx,
+		key:         key,
+		contentType: contentType,
+		opts:        opts,
+		urls: &serviceUrls{
+			serviceURL:   b.urls.serviceURL,
+			containerURL: &containerURL,
+			blockBlobURL: &blockBlobURL,
+		},
+		donec: make(chan struct{}),
+	}
+
+	return w, nil
+}
+
+// Write appends p to w. User must call Close to close the w after done writing.
+func (w *writer) Write(p []byte) (int, error) {
+	if w.w == nil {
+		if err := w.open(); err != nil {
+			return 0, err
+		}
+	}
+
+	select {
+	case <-w.donec:
+		return 0, w.err
+	default:
+	}
+
+	return w.w.Write(p)
+}
+
+func (w *writer) open() error {
+	pr, pw := io.Pipe()
+	w.w = pw
+
+	go func() {
+		defer close(w.donec)
+
+		var blobHTTPHeaders = azblob.BlobHTTPHeaders{}
+		if w.contentType != "" {
+			blobHTTPHeaders.ContentType = w.contentType
+		}
+		if len(w.opts.ContentMD5) > 0 {
+			blobHTTPHeaders.ContentMD5 = w.opts.ContentMD5
+		}
+
+		buf, err := ioutil.ReadAll(pr)
+		if err != nil {
+			w.err = err
+			pr.CloseWithError(err)
+			return
+		}
+
+		uploadOpts := azblob.UploadStreamToBlockBlobOptions{BufferSize: w.opts.BufferSize, MaxBuffers: defaultUploadBuffers, Metadata: w.opts.Metadata, BlobHTTPHeaders: blobHTTPHeaders}
+		_, err = azblob.UploadStreamToBlockBlob(w.ctx, bytes.NewReader(buf), *w.urls.blockBlobURL, uploadOpts)
+		w.err = err
+		
+		if err != nil {
+			w.err = err
+			pr.CloseWithError(err)
+			return
+		}
+	}()
+
+	return nil
+}
+
+// Close completes the writer and close it. Any error occuring during write will
+// be returned. If a writer is closed before any Write is called, Close will
+// create an empty file at the given key.
+func (w *writer) Close() error {
+
+	select {
```

Comment:
You should not have the `select` here, just remove it entirely from this function. It should block until the goroutine started in `open` finishes.

Reply:
Done